### PR TITLE
feat: build updater

### PR DESCRIPTION
**Pull request by LoneGit**

### CODE REVIEW

File: src/controller.py [hunk 1]
```diff
@@ -780,3 +781,11 @@ def update_interface(self, filename = ""):
...
         else:
             self.view.warning_nozzle_and_table_collision.setText("")
+
+    def open_updater(self):
+        subprocess.Popen("updater.exe")
+
+def save_splanes_to_file(splanes, filename):
+    with open(filename, 'w') as out:
+        for p in splanes:
+            out.write(p.toFile() + '\n')
```
Comment (lvjonok) [inline src/controller.py:791]: why we need this?

File: src/controller.py [hunk 2]
```diff
@@ -780,3 +781,11 @@ def update_interface(self, filename = ""):
             self.view.warning_nozzle_and_table_collision.setText(self.view.locale.WarningNozzleAndTableCollision + s.slicing.planes_contact_with_nozzle)
         else:
             self.view.warning_nozzle_and_table_collision.setText("")
+
+    def open_updater(self):
+        subprocess.Popen("updater.exe")
```
Comment (lvjonok) [inline src/controller.py:786]: try to use `./updater` because we will use this not only for windows

File: src/entry_window.py [hunk 3]
```diff
@@ -204,5 +207,36 @@ def open_existing_project(self):
...
+        project_settings_filename = str(pathlib.Path(project_path, settings_filename))
+
+        build_version = get_version(settings_filename)
+
+        try:
+            project_version = get_version(project_settings_filename)
+        except Exception as e:
+            project_version = ""
+
+        if (not project_version) or (build_version > project_version):
```
Comment (lvjonok) [inline src/entry_window.py:226]: once again you compare just strings

File: src/entry_window.py [hunk 4]
```diff
@@ -204,5 +207,36 @@ def open_existing_project(self):
...
         self.add_recent_project(selected_project)
 
+        self.сheck_project_version(selected_project)
+
         # emit signal with path to project file
         self.open_project_signal.emit(selected_project)
+
+    def сheck_project_version(self, project_path):
+        settings_filename = "settings.yaml"
+        project_settings_filename = str(pathlib.Path(project_path, settings_filename))
```
Comment (lvjonok) [inline src/entry_window.py:217]: I think we have `project_settings_filename` somewhere in `PathBuilder.<...>` better not to invent something here

File: src/entry_window.py [hunk 5]
```diff
@@ -204,5 +207,36 @@ def open_existing_project(self):
...
+            message_box.setText(locale.SettingsUpdate)
+            message_box.addButton(QMessageBox.Yes)
+            message_box.addButton(QMessageBox.No)
+            message_box.button(QMessageBox.Yes).setText(locale.Update)
+            message_box.button(QMessageBox.No).setText(locale.ContinueWithoutUpdating)
+
+            reply = message_box.exec()
+
+            if reply == QMessageBox.Yes:
+                project_settings_old_filename = str(pathlib.Path(project_path, "settings_old.yaml"))
```
Comment (lvjonok) [inline src/entry_window.py:239]: also add this to `PathBuilder`

File: src/settings.py [hunk 6]
```diff
@@ -93,6 +93,29 @@ def save_splanes_to_file(splanes, filename):
         for p in splanes:
             out.write(p.toFile() + '\n')
 
+def get_version(settings_filename):
+    with open(settings_filename, "r") as settings_file:
+        settings = yaml.safe_load(settings_file)
+
+    version = settings["common"]["version"]
```
Comment (lvjonok) [inline src/settings.py:100]: Better to use `get` for dictionaries, at least for `"version"` or wrap with try/except because it will be sad to crash here

File: src/settings.py [hunk 7]
```diff
@@ -93,6 +93,29 @@ def save_splanes_to_file(splanes, filename):
...
+    return version
+
+def paths_transfer_in_settings(initial_settings_filename, final_settings_filename):
+    with open(initial_settings_filename, "r") as settings_file:
+        initial_settings = yaml.safe_load(settings_file)
+
+    splanes_file = initial_settings["slicing"]["splanes_file"]
```
Comment (lvjonok) [inline src/settings.py:107]: we have this at `PathBuilder` class

File: src/settings.py [hunk 8]
```diff
@@ -94,24 +94,33 @@ def save_splanes_to_file(splanes, filename):
...
+    try:
+        with open(settings_filename, "r") as settings_file:
+            settings = yaml.safe_load(settings_file)
 
-    version = settings["common"]["version"]
-    return version
+        version = settings["common"]["version"]
+        return version
+    except Exception as e:
+        return ""
```
Comment (lvjonok) [inline src/settings.py:104]: ?) not even a message about what have happened?

File: src/bug_report.py [hunk 9]
```diff
@@ -79,6 +79,16 @@ def update_image_names_label(self):
...
         try:
+            error_description = self.error_description.toPlainText()
+
+            if not error_description:
+                message_box = QMessageBox(parent=self)
+                message_box.setWindowTitle(controller.view.locale.SubmittingBugReport)
+                message_box.setText(controller.view.locale.EmptyDescription)
+                message_box.setIcon(QMessageBox.Critical)
+                message_box.exec_()
+                return
```
Comment (l1va) [inline src/bug_report.py:90]: do not do like this. Bugs should go in a separate PR. [ok this time if you fix everything fast and we merge it in a few days]

File: src/entry_window.py [hunk 10]
```diff
@@ -204,5 +207,35 @@ def open_existing_project(self):
...
         self.open_project_signal.emit(selected_project)
+
+    def сheck_project_version(self, project_path):
+        sett().project_path = project_path
+        project_settings_filename = PathBuilder.settings_file()
+        build_version = sett().common.version
+
+        try:
+            project_version = get_version(project_settings_filename)
+        except Exception as e:
```
Comment (l1va) [inline src/entry_window.py:222]: why do you need try-except here if you have it inside the get_version ?

File: src/settings.py [hunk 11]
```diff
@@ -93,6 +93,39 @@ def save_splanes_to_file(splanes, filename):
...
+
+        version = settings["common"]["version"]
+        return version
+    except Exception as e:
+        print("Error reading version")
+        return ""
+
+def paths_transfer_in_settings(initial_settings_filename, final_settings_filename):
+    with open(initial_settings_filename, "r") as settings_file:
+        initial_settings = yaml.safe_load(settings_file)
```
Comment (l1va) [inline src/settings.py:109]: here you should update all settings, not only the paths. 
for k,v := in old_settings {
   new_settings[k]=v
}

File: src/locales.py [hunk 12]
```diff
@@ -120,6 +120,12 @@ class Locale:
     ChoosePrinterDirectory = "Choose printer name"
     AddNewPrinter = "Add new printer"
     DefaultPrinterWarn = "Be aware that you are using default printer. New data might be removed after update. We recommend to create new printer and calibrate it."
+    CheckUpdates = "Check for updates"
+    ProjectUpdate = "Project update"
+    SettingsUpdate = "The current project settings format is outdated. This may lead to errors in the program. Update format? When updating, the project settings will be reset!"
```
Comment (l1va) [inline src/locales.py:125]: They will not be reset if you copy all of them. Just mention that "some new settings will have default values. Please recheck them carefully".

**Review by l1va — CHANGES_REQUESTED** [review]
@lvjonok check my comments please and check that comments were fixed before the merge please.

File: src/locales.py [hunk 13]
```diff
@@ -120,6 +120,11 @@ class Locale:
     ChoosePrinterDirectory = "Choose printer name"
     AddNewPrinter = "Add new printer"
     DefaultPrinterWarn = "Be aware that you are using default printer. New data might be removed after update. We recommend to create new printer and calibrate it."
+    CheckUpdates = "Check for updates"
+    ProjectUpdate = "Project update"
+    SettingsUpdate = "The current project settings format is outdated. This may lead to errors in the program. Update format? When updating, some settings will be set to default!"
```
Comment (l1va) [inline src/locales.py:125]: "The current project's settings are outdated. This may lead to errors in the program. Do you want to update? New setting's fields will be set to default values, please recheck."

File: src/locales.py [hunk 14]
```diff
@@ -247,6 +252,11 @@ def __init__(self, **entries):
         ChoosePrinterDirectory = "Выберите название принтера",
         AddNewPrinter = "Добавить новый принтер",
         DefaultPrinterWarn = "Будьте внимательны, Вы используете принтер по умолчанию. Данные этого принтера будут перезаписываться при обновлениях. Мы рекомендуем создать и использовать свою конфигурацию принтера.",
+        CheckUpdates = "Проверить наличие обновлений",
+        ProjectUpdate = "Обновление проекта",
+        SettingsUpdate = "Формат настроек текущего проекта устарел. Это может привести к ошибкам в работе программы. Обновить формат? При обновлении некоторым настройкам будет установлено значение по умолчанию!",
```
Comment (l1va) [inline src/locales.py:257]: "Настройки текущего проекта устарели. Это может привести к ошибкам в работе программы. Хотите обновить? Новые поля настроек будут выставлены в значения по умолчанию, пожалуйста проверьте"

**Review by l1va — APPROVED** [review]
please update messages
